src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Query/PropertyPages/ProjectSetUIPropertyValueActionCore.cs

```diff
@@ -64,21 +65,39 @@ public async Task OnBeforeExecutingBatchAsync(IEnumerable<UnconfiguredProject> t
         {
             foreach (UnconfiguredProject project in targetProjects)
             {
-                if (!_rules.TryGetValue(project.FullPath, out List<IRule> projectRules))
+                if (!_rules.TryGetValue(project.FullPath, out List<IRule> projectRules)
+                    && await project.GetProjectLevelPropertyPagesCatalogAsync() is IPropertyPagesCatalog projectCatalog
+                    && projectCatalog.GetSchema(_pageName) is Rule rule
+                    && rule.GetProperty(_propertyName) is BaseProperty property)
                 {
+                    bool configurationDependent = property.IsConfigurationDependent();
                     projectRules = new List<IRule>();
-
                     IPropertyPageQueryCache propertyPageCache = _queryCacheProvider.CreateCache(project);
-                    if (await propertyPageCache.GetKnownConfigurationsAsync() is IImmutableSet<ProjectConfiguration> knownConfigurations)
+                    if (configurationDependent)
                     {
-                        foreach (ProjectConfiguration knownConfiguration in knownConfigurations.Where(config => config.MatchesDimensions(_dimensions)))
+                        // The property is configuration-dependent; we need to collect the bound rules for
+                        // all matching configurations.
+                        if (await propertyPageCache.GetKnownConfigurationsAsync() is IImmutableSet<ProjectConfiguration> knownConfigurations)
                         {
-                            if (await propertyPageCache.BindToRule(knownConfiguration, _pageName) is IRule boundRule)
+                            foreach (ProjectConfiguration knownConfiguration in knownConfigurations.Where(config => config.MatchesDimensions(_dimensions)))
                             {
-                                projectRules.Add(boundRule);
+                                if (await propertyPageCache.BindToRule(knownConfiguration, _pageName) is IRule boundRule)
```

Comment:
Since you have this 'filter' if-statement, you can either combine it with the Where, or (to save allocations) remove the Where and put an && condition in this if-statement for the `MatchesDimensions` clause. Either will remove double filtering.

Reply:
Done.